### ISSUE TITLE
[Typescript] types.array and types.map of (types.safeReference(Model)) should not believe it can take undefined values

### ISSUE THREAD

**Issue by dfee** [mobxjs/mobx-state-tree#1236]
**_Question_**

* [x] I've checked documentation and searched for existing issues
* [x] I tried the [spectrum channel](https://spectrum.chat/mobx-state-tree/general/usage-of-types-array-types-safereference-model~3ee219c7-c582-4c40-892d-a162e3f0d693)

As discussed in the Spectrum chat link above, I need to run the property that is created with `types.array(types.safeReference(Model))` through a filter function to remove undefined instances. Otherwise, this "view" into the data isn't compatible with something of type `Model[]`.

It's my understanding from the documentation that instance of `Model` will be removed from `types.array` if they're managed with `types.safeReference`. So I'm not sure why this filter function is necessary...

If I were to clone the list (i.e. `fooList = [...modelCollection.theArray]`) then certainly `fooList` could contain elements that are undefined (the referenced object could no longer exist). Is this why I need to filter it? Is this a bug with the typing of the safe reference MST array?

**Comment by xaviergonz** [comment]
In theory when the original (referenced) node is deleted from the tree, then the reference should be automatically removed from the parent array when using safeReference.

Can you post a minimal reproduction code?

**Comment by dfee** [comment]
```ts
import { Instance, SnapshotOrInstance, types } from "mobx-state-tree";

export const Foo = types.model("Foo", {
  id: types.identifier,
});

export const FooQuery = types
  .model("FooQuery", {
    id: types.identifier,
    items: types.array(types.safeReference(Foo)),
  })
  .actions(self => ({
    insert(ins: SnapshotOrInstance<typeof Foo>) {
      self.items.push(ins);
    },
  }));

export const FooCollection = types
  .model("FooCollection", {
    items: types.map(Foo),
    queries: types.map(FooQuery),
  })
  .actions(self => ({
    addQuery(ins: SnapshotOrInstance<typeof FooQuery>) {
      return self.queries.put(ins);
    },
    insert(ins: SnapshotOrInstance<typeof Foo>) {
      return self.items.put(ins);
    },
  }));

export const fooCollection = FooCollection.create();
export const fooQuery = fooCollection.addQuery({ id: "fq-0" });
export const foo = fooCollection.insert({ id: "f-0" });
fooQuery.insert(foo);

export const items: Instance<typeof Foo>[] = fooQuery.items; //err
```

```
Type 'IMSTArray<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>> & IStateTreeNode<IArrayType<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>>' is not assignable to type '(ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>)[]'.
  Types of property 'concat' are incompatible.
    Type '{ (...items: ConcatArray<(ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined>[]): ((ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined)[]; (...items: ConcatArray<string | number | (ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined>[]): ((ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined)[]; (...items: ((ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | ConcatArray<(ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined> | undefined)[]): ((ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined)[]; (...items: (string | number | (ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | ConcatArray<string | number | (ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined> | undefined)[]): ((ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined)[]; }' is not assignable to type '{ (...items: ConcatArray<ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>[]): (ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>)[]; (...items: ((ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>) | ConcatArray<ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>)[]): (ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>)[]; }'.
      Type '((ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined)[]' is not assignable to type '(ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>)[]'.
        Type '(ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IMaybe<IReferenceType<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>>>) | undefined' is not assignable to type 'ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>'.
          Type 'undefined' is not assignable to type 'ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>> & IStateTreeNode<IModelType<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>, {}, _NotCustomized, _NotCustomized>>'.
            Type 'undefined' is not assignable to type 'ModelInstanceTypeProps<ModelPropertiesDeclarationToProperties<{ id: ISimpleType<string>; }>>'.ts(2322)
```

**Comment by dfee** [comment]
For additional clarification, I'm not able to reproduce that there are `undefined` values in the array, only that either:
1) there should be `undefined` values in the array (and the documentation need to be updated)
2) there shouldn't be `undefined` values in the array (and the types need to be updated)
3) ... some other misunderstanding on my end :)

**Comment by xaviergonz** [comment]
That sounds actually like a typescript false positive

The thing is that the type of a safe reference (since it can be set to undefined at any moment) is either the object or undefined (which is right for model properties), but when the array takes it, it thinks it is an array of models or undefined values (which in runtime cannot ever happen)

Same goes for maps.

To fix it there should be a rule in typescript for arrays and maps exclusively for safe refs where it will strip out undefined from the possible values.

**Comment by xaviergonz** [comment]
btw, for now what you can do is

```
type NotMaybeType<IT extends IAnyType> = IT extends IMaybe<infer T> ? T : IT
function notMaybe<IT extends IAnyType>(type: IT): NotMaybeType<IT> {
    return type as any;
}

// in model props
safeRefArrayWithoutUndefined: types.array(notMaybe(types.safeReference(SomeModel)))
```

**Comment by xaviergonz** [comment]
Actually after some thought it is not actually a false positive since there are rare cases where an array of safe references can actually have undefined values:
* if the input snapshot has undefined values they will be kept (because it is a maybe type)
* if the array slot is manually set to undefined it is valid and kept (because it is a maybe type)

Although rare, it is possible, so the type should reflect that.

I just made a PR with a `safeReferenceForCollection` better suited for those cases

**Comment by mweststrate** [comment]
Released as 3.14.1
